drush

```diff
@@ -1,131 +1,350 @@
+# the "drush finder" script.
+#
+# The backend call will always set --root.  The "drush finder" script
+# always favors the site-local Drush stored with the site indicated by the
+# --root option, if it exists.  If there is no site-local Drush, then the
+# "drush finder" will behave as usual (i.e., it will end up calling the
+# "drush launcher" located next to it).
+#
+# This should always get you the correct "Drush" for local and remote calls.
+# Note that it is also okay for aliases to specify a path directly to
+# drush.launcher, in instances where it is known that a recent version of
+# Drush is installed on the remote end.
 #
 
+
 # Get the absolute path of this executable
 SELF_DIRNAME="`dirname -- "$0"`"
 SELF_PATH="`cd -P -- "$SELF_DIRNAME" && pwd -P`/`basename -- "$0"`"
 
-# Decide if we are running a Unix shell on Windows
-if `which uname > /dev/null 2>&1`; then
-  case "`uname -a`" in
-    CYGWIN*)
-      CYGWIN=1 ;;
-    MINGW*)
-      MINGW=1 ;;
-  esac
-fi
-
-# Resolve symlinks - this is the equivalent of "readlink -f", but also works with non-standard OS X readlink.
-while [ -h "$SELF_PATH" ]; do
-    # 1) cd to directory of the symlink
-    # 2) cd to the directory of where the symlink points
-    # 3) Get the pwd
-    # 4) Append the basename
-    DIR="`dirname -- "$SELF_PATH"`"
-    SYM="`readlink "$SELF_PATH"`"
-    SYM_DIRNAME="`dirname -- "$SYM"`"
-    SELF_PATH="`cd "$DIR" && cd "$SYM_DIRNAME" && pwd`/`basename -- "$SYM"`"
-done
-
-# If not exported, try to determine and export the number of columns.
-# We do not want to run `tput cols` if $TERM is empty, "unknown", or "dumb", because
-# if we do, tput will output an undesirable error message to stderr.  If
-# we redirect stderr in any way, e.g. `tput cols 2>/dev/null`, then the
-# error message is suppressed, but tput cols becomes confused about the
-# terminal and prints out the default value (80).
-if [ -z $COLUMNS ] && [ -n "$TERM" ] && [ "$TERM" != dumb ] && [ "$TERM" != unknown ] && [ -n "`which tput`" ] ; then
-  # Note to cygwin/mingw/msys users: install the ncurses package to get tput command.
-  # Note to mingw/msys users: there is no precompiled ncurses package.
-  if COLUMNS="`tput cols`"; then
-    export COLUMNS
-  fi
-fi
+FOUND_SCRIPT=
+SITE_USE_DIR="$HOME/.drush/use/$$/bin"
 
-if [ -n "$DRUSH_PHP" ] ; then
-  # Use the DRUSH_PHP environment variable if it is available.
-  php="$DRUSH_PHP"
-else
-  # On MSYSGIT, we need to use "php", not the full path to php
-  if [ -n "$MINGW" ] ; then
-    php="php"
-  else
-    # Default to using the php that we find on the PATH.
-    # We check for a command line (cli) version of php, and if found use that.
-    # Note that we need the full path to php here for Dreamhost, which behaves oddly.  See http://drupal.org/node/662926
-    php="`which php-cli 2>/dev/null`"
+# We will look for a "drush wrapper" script that might
+# be stored in the root of a Drupal site.  If there is
```

Comment:
We have been working towards moving references to Drupal to boot classes and commandfiles. Lets omit Drupal here.